### PR TITLE
HPCC-30835 Add esp-api command line tool for HPCC ESP services

### PR DESCRIPTION
**Pull request by eatekhan**
Introduce the esp-api command line tool with two main commands:
- `describe`: Describes all services, specific service methods, or the request and response structure of a method.
- `test`: Sends a request to a specified method.

Enhance usability with dynamic tab completion for service and method names:
- Provides real-time, context-sensitive suggestions to improve user experience and efficiency.

Usage and examples in the --help.

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [x] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Review by timothyklemm — COMMENTED** [review]
A number of comments made offline and discussed prior to the PR are awaiting updates.

File: tools/esp-api/espapi.cpp
Comment (asselitx) [inline tools/esp-api/espapi.cpp:108]: lowercase `Recursive`

File: tools/esp-api/espapi.cpp
Comment (asselitx) [inline tools/esp-api/espapi.cpp:120]: remove commented code

**Review by asselitx — CHANGES_REQUESTED** [review]
I've made some comments here for changes that I think are reasonable to make now. In the Word doc, I've looked through Tim's comments from Aug 5 and noted those that I think could be passed over and those that should be deferred to another ticket. Any items I didn't comment on I agree with.

In general if you're getting pressed for time I'd suggest deferring any disruptive changes to a future ticket.

File: esp/esdllib/esdl_def.cpp
Comment (timothyklemm) [inline esp/esdllib/esdl_def.cpp:1282]: If you insert multiple spaces between the type and name, as if to align a series of types and names, then the names should be aligned.

File: esp/esdllib/esdl_def.cpp
Comment (timothyklemm) [inline esp/esdllib/esdl_def.cpp:1294]: Whitespace... I personally would not put so much space between the method signature and the implementation, but with so much space present there is no reason for the implementation of `query` to be indented even more.

File: tools/esp-api/espapi.cpp
Comment (timothyklemm) [inline tools/esp-api/espapi.cpp:273]: The leading space, currently added unconditionally on line 262, should be included in this string. It should only be added if meta data is going to be output.

File: tools/esp-api/espapi.cpp
Comment (timothyklemm) [inline tools/esp-api/espapi.cpp:329]: See prior comment on the leading space.

File: tools/esp-api/espservice.cpp
Comment (timothyklemm) [inline tools/esp-api/espservice.cpp:39]: There is no use case where a password is added without a username that I am aware of. This condition should only be executed when a username is given.

File: tools/esp-api/espservice.cpp
Comment (timothyklemm) [inline tools/esp-api/espservice.cpp:59]: To be consistent with the error message on 65, consider `cerr << "Response code " << responseCode << ": Enter a valid esp target" << endl`

File: tools/esp-api/espservice.cpp
Comment (timothyklemm) [inline tools/esp-api/espservice.cpp:99]: Remove if unused.

File: tools/esp-api/espshell.cpp
Comment (timothyklemm) [inline tools/esp-api/espshell.cpp:34]: Replace comma with semicolon.

File: tools/esp-api/espshell.cpp
Comment (timothyklemm) [inline tools/esp-api/espshell.cpp:38]: Replace in with of.

File: tools/esp-api/espshell.cpp
Comment (timothyklemm) [inline tools/esp-api/espshell.cpp:70]: Be aware of redundant code. In this case, `url` can be set unconditionally up to and including `resType`. Then `"?" << query` can be added conditionally.

File: tools/esp-api/espshell.cpp
Comment (timothyklemm) [inline tools/esp-api/espshell.cpp:78]: Be aware of unnecessary data copies. You are creating local strings here and in `buildUrl`. When asking for a string to be populated, we usually pass the target string to the function by reference and fill it in. The buffer gets allocated and populated one time only.

**Review by timothyklemm — CHANGES_REQUESTED** [review]

File: tools/esp-api/espshell.cpp
Comment (asselitx) [reply tools/esp-api/espshell.cpp:70]: This may be a good candidate to defer to a 'cleanup' ticket if you don't have time to do it now. If so please post a link to the ticket in a comment.

File: tools/esp-api/espshell.cpp
Comment (asselitx) [reply tools/esp-api/espshell.cpp:78]: This may be a good candidate to defer to a 'cleanup' ticket if you don't have time to do it now. If so please post a link to the ticket in a comment.

**Review by asselitx — CHANGES_REQUESTED** [review]
This will be good to go once you address Tim's outstanding requests, and/or a link to the ticket you're deferring any to. Next steps are to commit & push those changes, give us a chance to quickly review, then once we both approve you can squash & push the final time.

File: tools/esp-api/espshell.cpp
Comment (timothyklemm) [inline tools/esp-api/espshell.cpp:24]: Unnecessary since you reverted to std::string. Not harmful,; doesn't need to be fixed.

**Review by timothyklemm — COMMENTED** [review]
If another commit is expected with a new readme, I'll just comment for now.

File: tools/esp-api/README.md
Comment (timothyklemm) [inline tools/esp-api/README.md:14]: Choose between:
1. Each command is invoked by `esp-api <command>`.
2. Invoked by the auto-complete script. The last sentence would then be redundant.

I would pick #2.

**Review by asselitx — APPROVED** [review]
looks good to squash

**Review by timothyklemm — APPROVED** [review]

**Comment by ghalliday** [comment]
@eatekhan this is currently failing the unit tests with the error:

1) test: EspApiCmdTest::testDescribeAllServices (F) line: 344 /hpcc-dev/HPCC-Platform/testing/unittests/espapicmdtests.cpp
assertion failed
- Expression: refString == res

**Comment by eatekhan** [comment]
> @eatekhan this is currently failing the unit tests with the error:
> 
> 1. test: EspApiCmdTest::testDescribeAllServices (F) line: 344 /hpcc-dev/HPCC-Platform/testing/unittests/espapicmdtests.cpp
>    assertion failed
> 
> * Expression: refString == res

Hmmm, I don't know why, I just reran them and all tests seem to pass on my side, can you provide some more information?
Additionally,  @timothyklemm or @asselitx , does it work on your side?

**Comment by timothyklemm** [comment]
> > @eatekhan this is currently failing the unit tests with the error:
> > 
> > 1. test: EspApiCmdTest::testDescribeAllServices (F) line: 344 /hpcc-dev/HPCC-Platform/testing/unittests/espapicmdtests.cpp
> >    assertion failed
> > 
> > 
> > * Expression: refString == res
> 
> Hmmm, I don't know why, I just reran them and all tests seem to pass on my side, can you provide some more information? Additionally, @timothyklemm or @asselitx , does it work on your side?

I can confirm that the tests pass in my bare-metal environment. I would suggest replacing the CPPUNIT_ASSERT calls (I would replace all of them, not just the current failure) with different variants such as CPPUNIT_ASSERT_EQUAL(expected, actual).  I believe the more specific assertion will actually print the values when the assertion fails.

**Comment by eatekhan** [comment]
Hey Guys, I made another commit to address the build failing in one of checks as recommended by Tim in his emails, can someone restart the checks, I don't think I have permissions to do so.

**Review by asselitx — APPROVED** [review]

**Review by timothyklemm — APPROVED** [review]

**Pull request merged**
